Move memory provider to core

nyalldawson opened the pull request:
This commit introduces basic support for non-library based data providers, and moves the memory provider from a library based provider to a core provider.

The memory provider is used so frequently throughout QGIS that it has become integral to QGIS functionality and must be available wherever the QGIS core library is used.

The changes here (while not exposed yet to Python) could potentially be built on to allow creation of pure Python data providers.

nyalldawson committed:
Move memory provider to core

This commit introduces basic support for non-library based
data providers, and moves the memory provider from a library
based provider to a core provider.

The memory provider is used so frequently throughout QGIS that
it has become integral to QGIS functionality and must be
available wherever the QGIS core library is used.

The changes here (while not exposed yet to Python) could potentially
be built on to allow creation of pure Python data providers.

nyalldawson committed:
Add QgsMemoryProviderUtils.createMemoryLayer for easy creation
of new memory layers from QgsFields/QgsCoordinateReferenceSystem/etc

Since memory provider cannot use QgsVectorLayerImport we need
another centeralized function for creating new memory layers
without the need to manually create the uri string.

nyalldawson committed:
Use QgsMemoryProviderUtils::createMemoryLayer when pasting as scratch layer

Also avoids annoying invalid "missing CRS" message/window when pasting
features to a scratch layer

nyalldawson commented:
@wonder-sk assigning to you as I know you've expressed interest in this in the past

nyalldawson committed:
Remove memory provider library from packaging

m-kuhn commented on src/core/CMakeLists.txt:
```diff
@@ -1000,6 +1009,8 @@ INCLUDE_DIRECTORIES(
   metadata
   pal
   processing
+  providers
```
Are you intending to move more providers here?

nyalldawson replied:
Yes, ogr and gdal would be the other candidates

m-kuhn replied:
So apparently memory provider can be used for custom python providers (though I can't really follow how exactly), that makes sense - although I'm not completely sold yet to the idea of (ab)using memory provider for this.

What's the advantage of integrating gdal/ogr in core?

nyalldawson replied:
No, the provider is not exposed through the bindings, so end use remains unchanged. What's changed here is the requirement that all providers are loaded via libraries. Now it's possible to register providers directly with function pointers. So this should be moderatly easy to expand and expose to the bindings, hence allowing pure python data providers.

m-kuhn replied:
I see, sounds good.

So what's the reason for shipping memory (and gdal/ogr) in core? Don't get me wrong, I don't see any downsides, just wondering if you've got some incredibly cool added value with shipping in core which I don't see.

nyalldawson replied:
There's a couple of reasons

My main reason was I needed the memory provider utils part. And then it seemed odd that we'd have utilities for working with certain providers in core when those providers themselves aren't in core.

Then it became an exercise in seeing what was missing from enabling plugin based providers...

m-kuhn commented on src/core/providers/memory/qgsmemoryproviderutils.cpp:
```diff
@@ -0,0 +1,83 @@
+{
+  QString geomType = QgsWkbTypes::displayString( geometryType );
+  if ( geomType.isNull() )
+    geomType = "none";
```
QStringLiteral?

nyalldawson replied:
Fixed

nyalldawson committed:
QStringLiteral

wonder-sk reviewed (approved):
The code looks great, the memory layer creation will be very useful!

Optional suggestions :-)
- It may be good to update docstrings of functions in QgsProviderRegistry (library(), providerLibrary(), function()) and QgsProviderMetadata (library()) to indicate that library will be null if it is internal provider.
- The code in QgsMemoryProviderUtils::createMemoryLayer() could be maybe simplified a little bit - to use a list for all key/value pairs (including CRS) and then joining those - to avoid need for the 'first' variable.

nyalldawson committed:
Update QgsProviderRegistry docs

nyalldawson committed:
Add missing /Factory/ annotations to QgsProviderMetadata

Also sipify QgsProviderMetadata and QgsProviderRegistry

nyalldawson committed:
Rename some QgsProviderRegistry methods to clarify that they are factories

- provider() was renamed to createProvider()
- selectWidget() was renamed to createSelectionWidget()
- providerLibrary() was renamed to createProviderLibrary()

nyalldawson committed:
Simplify QgsMemoryProviderUtils::createMemoryLayer()

nyalldawson commented:
@wonder-sk 

> It may be good to update docstrings of functions in QgsProviderRegistry (library(), providerLibrary(), function()) and QgsProviderMetadata (library()) to indicate that library will be null if it is internal provider.

Fixed

> The code in QgsMemoryProviderUtils::createMemoryLayer() could be maybe simplified a little bit - to use a list for all key/value pairs (including CRS) and then joining those - to avoid need for the 'first' variable.

Thanks, fixed
